First draft of GridLAB-D docker image

knyghty opened the pull request:
Ok, this is my first attempt.

So far, we can compile GridLAB-D with MySQL and HELICS support.

### To do
- [x] Some way of running with HELICS. Ideally, gridlabd and helics run in separate containers, and this repo only has the gridlabd part. However, we may need to have them both in the same container, and helics runs gridlabd directly. In this case, we rename this project `helics-gridlabd-docker` and go about our business.
- [ ] A better way of running config files than copying it into the container.
- [ ] Documentation on how to run this.

### Nice to have
- [ ] Alpine
- [x] Debian (at least bullseye)
- [ ] Multiple GridLAB-D versions - only really feasible without HELICS, otherwise we have a huge matrix of GridLAB-D / HELICS versions to maintain.
- [ ] CI tests (the test.glm file is a good basis).
- [ ] I tried to keep the image size as small as possible. Without MySQL and HELICS it was a reasonable ~170MB. With them, it's around 500MB. I'm not sure what more can be pruned, however.

### Notes

HELICS maintain docker images: https://github.com/GMLC-TDC/HELICS/tree/main/config/Docker - I haven't looked into them much, I don't know if they are useful. If so, we should use them if possible.

Ideally we have one container for GridLAB-D, one for helics, and one for the helics broker, though we may end up needing one for helics + gridlab-d, and one for the helics broker.

If helics' docker images don't suit our needs for whatever reason, we can maintain our own in another repo.

As I understand it, they communicate by ZeroMQ, which is a kind of brokerless message queue. It's possible to use a broker, and maybe this would be useful. I'm not sure if one is required when operating over the network or not.

We shouldn't have this test.glm file copied in like this, but as mentioned earlier, it could make a good basis for CI testing.

We shouldn't commit the `docker-compose.yml` but I've left it in this branch for now as it's useful for testing.

knyghty committed:
First draft

ryangilmour commented:
Not really clued up on what is required here, but I came across another project using gridlab-d and they have an open repo for a gridlabd docker image that might be useful? 

https://github.com/slacgismo/gridlabd-docker

devanubis commented:
@ryangilmour yeh they've got a huge fork of GridLab-D for which they use docker for development. [https://github.com/slacgismo/gridlabd](https://github.com/slacgismo/gridlabd) (it's diverges massively from the main project)

Unless we want to move to using their fork I don't thin we'll be able to use their docker images. But they may be helpful in getting our image to work with Helics in a separate container (if that's even possible).

devanubis committed:
Second draft

Using debian:bullseye-slim as a base, a multi-stage build to make things
a little easier to read and debug, and a test docker-compose.yml which
spins up a mysql server and helics_broker to connect to.

devanubis commented:
@knyghty  I've pushed my changes for a second draft of a gridlab-d image.

- debian:bullseye-slim
- helics 2.7.1
  - The latest compatible with gridlabd
  - https://github.com/gridlab-d/gridlab-d/issues/1279
- gridlabd v4.3
  - Using our fork, with a minor patch for mariadb compatibility
- Final image is `436MB`
  - I used a multi-stage build as it made debugging each compile step a bit easier, and the intermediary images do get discarded during the build)
  - I'm happy to move back to a more convoluted build though if you think it could optimise the image size.
  - I could also base our `gridlabd` image on of the `helics` image and re-cycling that to get pre-compiled helics binaries... Helics is a pretty huge chunk of the build time...

knyghty reviewed (commented):
(I cannot approve since I started the PR)

It looks good to me... We should also add a CI workflow to build the image and push it to GHCR so we can use it in docker compose without having to clone this repo somewhere.

devanubis commented:
I was just going to use it as:  
```yaml
build:  
  context: https://github.com/scene-connect/gridlab-d-docker.git#draft
```
  
But pre-building images and pulling sounds much nicer 👍‍

devanubis committed:
Adds CI step to build and push container image

knyghty commented on .github/workflows/cy.yml:
```diff
@@ -0,0 +1,29 @@
+  push:
+    branches:
+      - main
+  pull_request:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  push_to_ghcr:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```
If this is the only job then we could remove `pull_request:` and the `if` here and it'll only run on pushes to main.

I'd suggest giving it a test as well without this so we can see it working.

knyghty commented on .github/workflows/cy.yml:
```diff
@@ -0,0 +1,28 @@
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and push
+        uses: docker/build-push-action@v2
+        with:
```
Once the image is pushed we can add `cache-from` here to use cached layers from the previously pushed image.

It can also be a good idea to tag and push the builder image so that can be used as a cache, too.

devanubis commented:
OK I've got it building the package and hitting the caches (finally!)  
  
I'll set the same up for the simulation image next.
